PythonVisualizations/SortingBase.py

```diff
@@ -0,0 +1,420 @@
+        self.canvas.lower(rect)
+        return rect        
+    
+    def createCellValue(self, indexOrCoords, key, color=None):
+        """Create new canvas items to represent a cell value.  A square
+        is created filled with a particular color with a text key centered
+        inside.  The position of the cell can either be an integer index in
+        the Array or the bounding box coordinates of the square.  If color
+        is not supplied, the next color in the palette is used.
+        An event handler is set up to update the VisualizationApp argument
+        with the cell's value if clicked with any button.
+        Returns the tuple, (square, text), of canvas items
+        """
+        # Determine position and color of cell
+        if isinstance(indexOrCoords, int):
+            rectPos = self.cellCoords(indexOrCoords)
+            valPos = self.cellCenter(indexOrCoords)
+        else:
+            rectPos = indexOrCoords
+            valPos = divide_vector(add_vector(rectPos[:2], rectPos[2:]), 2)
+        if color is None:
+            # Take the next color from the palette
+            color = drawable.palette[SortingBase.nextColor]
+            SortingBase.nextColor = (SortingBase.nextColor + 1) % len(drawable.palette)
+      
+        newCoords = self.fillCoords(key, rectPos)
+    
+        cell_rect = self.canvas.create_rectangle(
+            *newCoords, fill=color, outline='', width=0)
+        cell_val = self.canvas.create_text(
+            *valPos, text=str(key), font=self.VALUE_FONT, fill=self.VALUE_COLOR, tags="cellVal")
+        handler = lambda e: self.setArgument(str(key))
+        for item in (cell_rect, cell_val):
+            self.canvas.tag_bind(item, '<Button>', handler)
+
+        return cell_rect, cell_val
+    
+    # get the whole box coords and return how much will get filled 
+    def fillCoords(self, val, rectPos): 
+        x1, y1, x2, y2 = rectPos 
+        midY = (y1 + y2) // 2
+        # proportion of what is filled
+        prop = val / 99
+        y2 = round(midY + prop * (y2-midY))
+        
+        return (x1, y1, x2, y2)
+    
+    def redrawArrayCells(self):        
+        self.canvas.delete("arrayBox")
+        for i in range(len(self.list)):
+            self.createArrayCell(i)    
+    
+    def display(self):
+        self.canvas.delete("all")
+    
+        for i in range(self.size):  # Draw grid of cells
+            self.createArrayCell(i)
+    
+        # go through each Drawable in the list
+        for i, n in enumerate(self.list):
+            # create display objects for the associated Drawables
+            n.display_shape, n.display_val = self.createCellValue(
+                i, n.val, n.color)
+            if self.changeSize:
+                self.canvas.delete(n.display_val)
+            n.color = self.canvas.itemconfigure(n.display_shape, 'fill')
+    
+        self.window.update()
```

Comment:
The `display()` method creates the canvas items to display the core of the data structure, an array of a particular `size` with a `nItems` index.  The `nItems` index should be created here and moved around by the subclasses.  This will be more consistent with the `nItems` indices added via PR #113.

The loop on line 284 should only go through the first `nItems` of `self.list`.  By going through the allocated size, it might encounter cells with no values.

Line 290 needs a `[-1]` at the end to extract the last element of the tuple returned by the call to `itemconfigure()`.  Otherwise the whole tuple will be stored in the drawable's `color` field.

Reply:
Since PR #113 was merged into `master`, it probably makes sense to merge `master` into this development branch to get the changes that make the `nItems` pointer a "permanent" index on the canvas.

Reply:
Fixed common `nItems` pointer in commit 4ae34f2.  I fixed the other problems, too.